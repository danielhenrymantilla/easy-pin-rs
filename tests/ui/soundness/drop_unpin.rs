#![feature(trivial_bounds)]
#![allow(unused)]

use ::easy_pin::{
    easy_pin,
    PinDrop,
    PinSensitive,
};
use ::core::{
    pin::Pin,
    ptr::NonNull,
};

#[easy_pin(Drop)]
pub
struct SelfReferential {
    #[transitively_pinned]
    string: PinSensitive<String>,

    #[unpinned]
    at_string: NonNull<String>,
}

impl PinDrop for SelfReferential {
    unsafe
    fn drop_pinned (mut self: Pin<&mut Self>)
    {
        let at_string_mut: &mut PinSensitive<String> = if true {
            self.pinned_string_mut().get_mut()
        } else {
            &mut self.get_mut().string
        };
        ::core::mem::replace(
            at_string_mut,
            PinSensitive::new("".into()),
        );
    }
}

fn main ()
{}
