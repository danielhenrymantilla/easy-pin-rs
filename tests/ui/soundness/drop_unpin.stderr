error[E0277]: the trait bound `std::marker::PhantomPinned: std::marker::Unpin` is not satisfied in `easy_pin::PinSensitive<std::string::String>`
  --> $DIR/drop_unpin.rs:29:38
   |
29 |             self.pinned_string_mut().get_mut()
   |                                      ^^^^^^^ within `easy_pin::PinSensitive<std::string::String>`, the trait `std::marker::Unpin` is not implemented for `std::marker::PhantomPinned`
   |
   = help: the following implementations were found:
             <std::marker::PhantomPinned as std::marker::Unpin>
   = note: required because it appears within the type `easy_pin::PinSensitive<std::string::String>`

error[E0277]: the trait bound `std::marker::PhantomPinned: std::marker::Unpin` is not satisfied in `easy_pin::PinSensitive<std::string::String>`
  --> $DIR/drop_unpin.rs:31:23
   |
31 |             &mut self.get_mut().string
   |                       ^^^^^^^ within `easy_pin::PinSensitive<std::string::String>`, the trait `std::marker::Unpin` is not implemented for `std::marker::PhantomPinned`
   |
   = help: the following implementations were found:
             <std::marker::PhantomPinned as std::marker::Unpin>
   = note: required because it appears within the type `easy_pin::PinSensitive<std::string::String>`
   = note: required because of the requirements on the impl of `std::marker::Unpin` for `SelfReferential`

For more information about this error, try `rustc --explain E0277`.
